raft/raft_test.go

```diff
@@ -1911,6 +1911,244 @@ func TestCommitAfterRemoveNode(t *testing.T) {
 	}
 }
 
+func TestNormalLeaderTransfer(t *testing.T) {
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	lead := nt.peers[1].(*raft)
+
+	if lead.lead != 1 {
+		t.Fatalf("After election leader is %x, want 1", lead.lead)
+	}
+
+	// Transfer leader to 2.
+	nt.send(pb.Message{From: 2, To: 1, Type: pb.MsgTransferLeader})
+
+	if lead.lead != 2 || lead.transferee != None {
+		t.Fatalf("After first time transferring, leader is %x transferee %v, want leader 2, transferee %v", lead.lead, lead.transferee, None)
+	}
+
+	// After some log replication, transfer leader back to 1.
+	for j := 0; j <= 5; j++ {
+		nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgProp, Entries: []pb.Entry{{}}})
+	}
+
+	nt.send(pb.Message{From: 1, To: 2, Type: pb.MsgTransferLeader})
+
+	if lead.lead != 1 || lead.transferee != None {
+		t.Fatalf("After second time transferring, leader is %x transferee %v, want leader 1, transferee %v", lead.lead, lead.transferee, None)
+	}
+}
+
+func TestTransferAfterLogReplication(t *testing.T) {
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	nt.isolate(3)
+	for j := 0; j <= 5; j++ {
+		nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgProp, Entries: []pb.Entry{{}}})
+	}
+
+	nt.recover()
+	lead := nt.peers[1].(*raft)
+	if lead.prs[3].Match != 1 {
+		t.Fatalf("Node 3 has match %x, want %x", lead.prs[3].Match, 1)
+	}
+
+	// Transfer leader to 3 when node 3 is lack of log.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+
+	if lead.lead != 3 || lead.transferee != None {
+		t.Fatalf("After transferring, leader is %x transferee %v, want leader 3, transferee %v", lead.lead, lead.transferee, None)
+	}
+}
+
+func TestTransferAfterSnapshot(t *testing.T) {
+	nt := newNetwork(nil, nil, nil)
+	nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgHup})
+
+	nt.isolate(3)
+	for j := 0; j <= 5; j++ {
+		nt.send(pb.Message{From: 1, To: 1, Type: pb.MsgProp, Entries: []pb.Entry{{}}})
+	}
+	lead := nt.peers[1].(*raft)
+	nextEnts(lead, nt.storage[1])
+	nt.storage[1].CreateSnapshot(lead.raftLog.applied, &pb.ConfState{Nodes: lead.nodes()}, nil)
+	nt.storage[1].Compact(lead.raftLog.applied)
+
+	nt.recover()
+	if lead.prs[3].Match != 1 {
+		t.Fatalf("Node 3 has match %x, want %x", lead.prs[3].Match, 1)
+	}
+
+	// Transfer leader to 3 when node 3 is lack of snapshot.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgTransferLeader})
+	// Only pb.MsgHeartbeatResp can trigger a snapshot.
+	nt.send(pb.Message{From: 3, To: 1, Type: pb.MsgHeartbeatResp})
+
+	if lead.lead != 3 || lead.transferee != None {
+		t.Fatalf("After transferring, leader is %x transferee %v, want leader 3, transferee %v", lead.lead, lead.transferee, None)
+	}
+}
+
+func TestTransferToMyselfOrNonExistNode(t *testing.T) {
```

Comment:
One test func for one test case? 

So TestLeaderTransferToSelf and TestLeaderTransferToNonExistingNode